File: src/diy_planning/diy_planning/a_star_planner_node.py
#!/usr/bin/env python3

"""
Fast A* planner for the DIY Challenge robot.

Consumes:
    /map
    /goal_pose
    TF map -> base_frame

Publishes:
    /a_star/path
    /a_star/visited_map

Features:
    - 8-connected A*
    - heapq priority queue
    - octile heuristic
    - precomputed obstacle inflation
    - O(1) safety lookup during planning
    - diagonal corner-cut prevention
    - visited-map visualization
    - transient-local QoS for map-like topics
"""

import heapq
import math

import rclpy
from geometry_msgs.msg import Pose, PoseStamped
from nav_msgs.msg import OccupancyGrid, Path
from rclpy.node import Node
from rclpy.qos import (
    DurabilityPolicy,
    QoSProfile,
    ReliabilityPolicy,
)
from tf2_ros import Buffer, TransformListener


class GraphNode:

    def __init__(
        self,
        x,
        y,
        cost=0.0,
        heuristic=0.0,
        prev=None
    ):
        self.x = x
        self.y = y
        self.cost = cost
        self.heuristic = heuristic
        self.prev = prev

    @property
    def total_cost(self):
        return self.cost + self.heuristic


class AStarPlanner(Node):

    def __init__(self):

        super().__init__('a_star_node')

        # ============================================================
        # TF
        # ============================================================

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(
            self.tf_buffer,
            self
        )

        # ============================================================
        # Parameters
        # ============================================================

        self.declare_parameter(
            'map_topic',
            '/map'
        )

        self.declare_parameter(
            'goal_topic',
            '/goal_pose'
        )

        self.declare_parameter(
            'path_topic',
            '/a_star/path'
        )

        self.declare_parameter(
            'visited_map_topic',
            '/a_star/visited_map'
        )

        self.declare_parameter(
            'base_frame',
            'base_footprint'
        )

        self.declare_parameter(
            'robot_clearance',
            0.40
        )

        self.declare_parameter(
            'occupied_threshold',
            50
        )

        self.declare_parameter(
            'unknown_is_occupied',
            True
        )

        self.declare_parameter(
            'visited_publish_interval',
            250
        )

        # ============================================================
        # Read parameters
        # ============================================================

        self.map_topic = self.get_parameter(
            'map_topic'
        ).value

        self.goal_topic = self.get_parameter(
            'goal_topic'
        ).value

        self.path_topic = self.get_parameter(
            'path_topic'
        ).value

        self.visited_map_topic = self.get_parameter(
            'visited_map_topic'
        ).value

        self.base_frame = self.get_parameter(
            'base_frame'
        ).value

        self.robot_clearance = float(
            self.get_parameter(
                'robot_clearance'
            ).value
        )

        self.occupied_threshold = int(
            self.get_parameter(
                'occupied_threshold'
            ).value
        )

        self.unknown_is_occupied = bool(
            self.get_parameter(
                'unknown_is_occupied'
            ).value
        )

        self.visited_publish_interval = int(
            self.get_parameter(
                'visited_publish_interval'
            ).value
        )

        if self.visited_publish_interval < 1:
            self.visited_publish_interval = 1

        # ============================================================
        # QoS
        # ============================================================

        self.map_qos = QoSProfile(
            depth=1
        )

        self.map_qos.reliability = (
            ReliabilityPolicy.RELIABLE
        )

        self.map_qos.durability = (
            DurabilityPolicy.TRANSIENT_LOCAL
        )

        # ============================================================
        # ROS interfaces
        # ============================================================

        self.map_sub = self.create_subscription(
            OccupancyGrid,
            self.map_topic,
            self.map_callback,
            self.map_qos
        )

        self.goal_sub = self.create_subscription(
            PoseStamped,
            self.goal_topic,
            self.goal_callback,
            10
        )

        self.path_pub = self.create_publisher(
            Path,
            self.path_topic,
            self.map_qos
        )

        self.visited_map_pub = self.create_publisher(
            OccupancyGrid,
            self.visited_map_topic,
            self.map_qos
        )

        # ============================================================
        # State
        # ============================================================

        self.map_ = None
        self.safe_grid = None

        self.visited_map_ = (
            OccupancyGrid()
        )

        self.clearance_cells = 0

        # heapq needs a deterministic tie breaker
        self.heap_counter = 0

        self.get_logger().info(
            'Fast A* planner ready'
        )

        self.get_logger().info(
            f'  robot_clearance = '
            f'{self.robot_clearance:.2f} m'
        )

    # ================================================================
    # Map callback
    # ================================================================

    def map_callback(
        self,
        map_msg: OccupancyGrid
    ):

        self.map_ = map_msg

        self.clearance_cells = int(
            math.ceil(
                self.robot_clearance
                / map_msg.info.resolution
            )
        )

        self.get_logger().info(
            f'Map received: '
            f'{map_msg.info.width} x '
            f'{map_msg.info.height}, '
            f'resolution='
            f'{map_msg.info.resolution:.3f} m'
        )

        self.get_logger().info(
            f'Inflating obstacles by '
            f'{self.robot_clearance:.2f} m '
            f'({self.clearance_cells} cells)'
        )

        # ------------------------------------------------------------
        # PRECOMPUTE SAFE GRID
        # ------------------------------------------------------------

        self.build_safe_grid()

        # ------------------------------------------------------------
        # Prepare visited map
        # ------------------------------------------------------------

        self.reset_visited_map()

        self.get_logger().info(
            'Inflated safety grid ready'
        )

    # ================================================================
    # Precompute inflated obstacle map
    # ================================================================

    def build_safe_grid(self):

        width = self.map_.info.width
        height = self.map_.info.height

        total_cells = (
            width * height
        )

        # ------------------------------------------------------------
        # Start by assuming every cell is safe.
        # ------------------------------------------------------------

        self.safe_grid = [
            True
        ] * total_cells

        # ------------------------------------------------------------
        # First mark original blocked cells.
        # ------------------------------------------------------------

        blocked_cells = []

        for y in range(height):

            row_offset = y * width

            for x in range(width):

                index = (
                    row_offset + x
                )

                value = (
                    self.map_.data[index]
                )

                blocked = False

                # Unknown
                if value < 0:

                    if self.unknown_is_occupied:
                        blocked = True

                # Occupied
                elif (
                    value
                    >= self.occupied_threshold
                ):

                    blocked = True

                if blocked:

                    self.safe_grid[index] = False

                    blocked_cells.append(
                        (x, y)
                    )

        # ------------------------------------------------------------
        # Precompute circular inflation offsets once.
        # ------------------------------------------------------------

        radius = (
            self.clearance_cells
        )

        inflation_offsets = []

        radius_squared = (
            radius * radius
        )

        for dy in range(
            -radius,
            radius + 1
        ):

            for dx in range(
                -radius,
                radius + 1
            ):

                if (
                    dx * dx
                    + dy * dy
                    <= radius_squared
                ):

                    inflation_offsets.append(
                        (dx, dy)
                    )

        # ------------------------------------------------------------
        # Inflate every blocked cell.
        # ------------------------------------------------------------

        for (
            obstacle_x,
            obstacle_y
        ) in blocked_cells:

            for (
                dx,
                dy
            ) in inflation_offsets:

                nx = (
                    obstacle_x + dx
                )

                ny = (
                    obstacle_y + dy
                )

                if (
                    0 <= nx < width
                    and
                    0 <= ny < height
                ):

                    index = (
                        ny * width
                        + nx
                    )

                    self.safe_grid[index] = False

        # ------------------------------------------------------------
        # Treat borders inside clearance radius as unsafe.
        #
        # This is equivalent to considering outside-map space blocked.
        # ------------------------------------------------------------

        if radius > 0:

            for y in range(height):

                for x in range(width):

                    if (
                        x < radius
                        or
                        y < radius
                        or
                        x >= width - radius
                        or
                        y >= height - radius
                    ):

                        self.safe_grid[
                            y * width + x
                        ] = False

        safe_count = sum(
            1
            for value in self.safe_grid
            if value
        )

        self.get_logger().info(
            f'Safe cells: '
            f'{safe_count}/{total_cells}'
        )

    # ================================================================
    # Goal callback
    # ================================================================

    def goal_callback(
        self,
        goal_msg: PoseStamped
    ):

        if self.map_ is None:

            self.get_logger().error(
                'No map received!'
            )

            return

        if self.safe_grid is None:

            self.get_logger().error(
                'Safety grid has not been generated.'
            )

            return

        # ------------------------------------------------------------
        # Frame check
        # ------------------------------------------------------------

        if (
            goal_msg.header.frame_id
            != self.map_.header.frame_id
        ):

            self.get_logger().error(
                f"Goal frame "
                f"'{goal_msg.header.frame_id}' "
                f"does not match map frame "
                f"'{self.map_.header.frame_id}'"
            )

            return

        # ------------------------------------------------------------
        # Reset visited visualization
        # ------------------------------------------------------------

        self.reset_visited_map()

        # ------------------------------------------------------------
        # Get robot location
        # ------------------------------------------------------------

        try:

            map_to_base_tf = (
                self.tf_buffer.lookup_transform(
                    self.map_.header.frame_id,
                    self.base_frame,
                    rclpy.time.Time()
                )
            )

        except Exception as exception:

            self.get_logger().error(
                f'Could not transform '
                f'{self.map_.header.frame_id} -> '
                f'{self.base_frame}: '
                f'{exception}'
            )

            return

        start_pose = Pose()

        start_pose.position.x = (
            map_to_base_tf.transform.translation.x
        )

        start_pose.position.y = (
            map_to_base_tf.transform.translation.y
        )

        start_pose.position.z = (
            map_to_base_tf.transform.translation.z
        )

        start_pose.orientation = (
            map_to_base_tf.transform.rotation
        )

        self.get_logger().info(
            f'Planning: '
            f'({start_pose.position.x:.2f}, '
            f'{start_pose.position.y:.2f}) '
            f'-> '
            f'({goal_msg.pose.position.x:.2f}, '
            f'{goal_msg.pose.position.y:.2f})'
        )

        path = self.plan(
            start_pose,
            goal_msg.pose
        )

        if path.poses:

            self.get_logger().info(
                f'Path found: '
                f'{len(path.poses)} poses'
            )

            self.path_pub.publish(
                path
            )

        else:

            self.get_logger().warn(
                'No safe path found.'
            )

    # ================================================================
    # Fast A*
    # ================================================================

    def plan(
        self,
        start: Pose,
        goal: Pose
    ) -> Path:

        start_node = (
            self.world_to_grid(
                start
            )
        )

        goal_node = (
            self.world_to_grid(
                goal
            )
        )

        path = Path()

        path.header.frame_id = (
            self.map_.header.frame_id
        )

        path.header.stamp = (
            self.get_clock()
            .now()
            .to_msg()
        )

        # ------------------------------------------------------------
        # Validate
        # ------------------------------------------------------------

        if not self.pose_on_map(
            start_node
        ):

            self.get_logger().error(
                'Start outside map'
            )

            return path

        if not self.pose_on_map(
            goal_node
        ):

            self.get_logger().error(
                'Goal outside map'
            )

            return path

        if not self.is_safe(
            start_node
        ):

            self.get_logger().error(
                'Start is inside inflated obstacle region'
            )

            return path

        if not self.is_safe(
            goal_node
        ):

            self.get_logger().error(
                'Goal is inside inflated obstacle region'
            )

            return path

        # ============================================================
        # Directions
        # ============================================================

        directions = [

            (-1, 0, 1.0),
            (1, 0, 1.0),
            (0, -1, 1.0),
            (0, 1, 1.0),

            (-1, -1, math.sqrt(2.0)),
            (-1, 1, math.sqrt(2.0)),
            (1, -1, math.sqrt(2.0)),
            (1, 1, math.sqrt(2.0)),
        ]

        # ============================================================
        # heapq open set
        # ============================================================

        open_heap = []

        start_key = (
            start_node.x,
            start_node.y
        )

        goal_key = (
            goal_node.x,
            goal_node.y
        )

        start_h = (
            self.octile_distance(
                start_node,
                goal_node
            )
        )

        # g_score dictionary
        g_score = {
            start_key: 0.0
        }

        # Parent dictionary:
        #
        # child -> parent
        came_from = {}

        self.heap_counter = 0

        heapq.heappush(
            open_heap,
            (
                start_h,
                self.heap_counter,
                start_node.x,
                start_node.y
            )
        )

        closed_set = set()

        expanded_count = 0

        goal_reached = False

        # ============================================================
        # Search
        # ============================================================

        while (
            open_heap
            and rclpy.ok()
        ):

            (
                current_f,
                _,
                current_x,
                current_y
            ) = heapq.heappop(
                open_heap
            )

            current_key = (
                current_x,
                current_y
            )

            if (
                current_key
                in closed_set
            ):

                continue

            closed_set.add(
                current_key
            )

            expanded_count += 1

            # --------------------------------------------------------
            # Visited map
            # --------------------------------------------------------

            current_index = (
                current_y
                * self.map_.info.width
                + current_x
            )

            original_value = (
                self.map_.data[
                    current_index
                ]
            )

            if (
                original_value
                < self.occupied_threshold
            ):

                self.visited_map_.data[
                    current_index
                ] = 50

            if (
                expanded_count
                % self.visited_publish_interval
                == 0
            ):

                self.publish_visited_map()

            # --------------------------------------------------------
            # Goal
            # --------------------------------------------------------

            if (
                current_key
                == goal_key
            ):

                goal_reached = True
                break

            current_g = (
                g_score[current_key]
            )

            # ========================================================
            # Neighbors
            # ========================================================

            for (
                dx,
                dy,
                movement_cost
            ) in directions:

                nx = (
                    current_x + dx
                )

                ny = (
                    current_y + dy
                )

                neighbor = (
                    GraphNode(
                        nx,
                        ny
                    )
                )

                if not self.pose_on_map(
                    neighbor
                ):

                    continue

                # O(1) lookup now
                if not self.is_safe(
                    neighbor
                ):

                    continue

                # ----------------------------------------------------
                # Diagonal corner cutting prevention
                # ----------------------------------------------------

                diagonal = (
                    dx != 0
                    and dy != 0
                )

                if diagonal:

                    horizontal = (
                        GraphNode(
                            current_x + dx,
                            current_y
                        )
                    )

                    vertical = (
                        GraphNode(
                            current_x,
                            current_y + dy
                        )
                    )

                    if (
                        not self.is_safe(
                            horizontal
                        )
                        or
                        not self.is_safe(
                            vertical
                        )
                    ):

                        continue

                neighbor_key = (
                    nx,
                    ny
                )

                tentative_g = (
                    current_g
                    + movement_cost
                )

                existing_g = (
                    g_score.get(
                        neighbor_key,
                        float('inf')
                    )
                )

                if (
                    tentative_g
                    >= existing_g
                ):

                    continue

                # ----------------------------------------------------
                # Better path
                # ----------------------------------------------------

                came_from[
                    neighbor_key
                ] = current_key

                g_score[
                    neighbor_key
                ] = tentative_g

                heuristic = (
                    self.octile_xy(
                        nx,
                        ny,
                        goal_node.x,
                        goal_node.y
                    )
                )

                f_score = (
                    tentative_g
                    + heuristic
                )

                self.heap_counter += 1

                heapq.heappush(
                    open_heap,
                    (
                        f_score,
                        self.heap_counter,
                        nx,
                        ny
                    )
                )

        # ============================================================
        # Final visited visualization
        # ============================================================

        self.publish_visited_map()

        self.get_logger().info(
            f'A* expanded '
            f'{expanded_count} cells'
        )

        if not goal_reached:

            return path

        # ============================================================
        # Reconstruct path
        # ============================================================

        cells = []

        current_key = (
            goal_key
        )

        cells.append(
            current_key
        )

        while (
            current_key
            != start_key
        ):

            if (
                current_key
                not in came_from
            ):

                self.get_logger().error(
                    'Broken A* parent chain'
                )

                return Path()

            current_key = (
                came_from[
                    current_key
                ]
            )

            cells.append(
                current_key
            )

        cells.reverse()

        # ------------------------------------------------------------
        # Convert cells -> Path
        # ------------------------------------------------------------

        for (
            x,
            y
        ) in cells:

            pose = (
                self.grid_to_world_xy(
                    x,
                    y
                )
            )

            pose_stamped = (
                PoseStamped()
            )

            pose_stamped.header = (
                path.header
            )

            pose_stamped.pose = pose

            path.poses.append(
                pose_stamped
            )

        return path

    # ================================================================
    # FAST safety lookup
    # ================================================================

    def is_safe(
        self,
        node: GraphNode
    ):

        if not self.pose_on_map(
            node
        ):

            return False

        index = (
            node.y
            * self.map_.info.width
            + node.x
        )

        return (
            self.safe_grid[index]
        )

    # ================================================================
    # Visited map
    # ================================================================

    def reset_visited_map(self):

        if self.map_ is None:
            return

        self.visited_map_ = (
            OccupancyGrid()
        )

        self.visited_map_.header = (
            self.map_.header
        )

        self.visited_map_.info = (
            self.map_.info
        )

        self.visited_map_.data = (
            list(self.map_.data)
        )

        self.publish_visited_map()

    def publish_visited_map(self):

        if self.map_ is None:
            return

        self.visited_map_.header.frame_id = (
            self.map_.header.frame_id
        )

        self.visited_map_.header.stamp = (
            self.get_clock()
            .now()
            .to_msg()
        )

        self.visited_map_pub.publish(
            self.visited_map_
        )

    # ================================================================
    # Heuristic
    # ================================================================

    def octile_distance(
        self,
        node,
        goal
    ):

        return self.octile_xy(
            node.x,
            node.y,
            goal.x,
            goal.y
        )

    def octile_xy(
        self,
        x1,
        y1,
        x2,
        y2
    ):

        dx = abs(
            x1 - x2
        )

        dy = abs(
            y1 - y2
        )

        minimum = min(
            dx,
            dy
        )

        maximum = max(
            dx,
            dy
        )

        return (
            maximum
            + (
                math.sqrt(2.0)
                - 1.0
            )
            * minimum
        )

    # ================================================================
    # Map helpers
    # ================================================================

    def pose_on_map(
        self,
        node
    ):

        return (
            0 <= node.x
            < self.map_.info.width
            and
            0 <= node.y
            < self.map_.info.height
        )

    def world_to_grid(
        self,
        pose: Pose
    ):

        x = int(
            math.floor(
                (
                    pose.position.x
                    - self.map_.info.origin.position.x
                )
                / self.map_.info.resolution
            )
        )

        y = int(
            math.floor(
                (
                    pose.position.y
                    - self.map_.info.origin.position.y
                )
                / self.map_.info.resolution
            )
        )

        return GraphNode(
            x,
            y
        )

    def grid_to_world_xy(
        self,
        x,
        y
    ):

        pose = Pose()

        pose.position.x = (
            (
                x + 0.5
            )
            * self.map_.info.resolution
            + self.map_.info.origin.position.x
        )

        pose.position.y = (
            (
                y + 0.5
            )
            * self.map_.info.resolution
            + self.map_.info.origin.position.y
        )

        pose.position.z = 0.0

        pose.orientation.x = 0.0
        pose.orientation.y = 0.0
        pose.orientation.z = 0.0
        pose.orientation.w = 1.0

        return pose


def main(args=None):

    rclpy.init(
        args=args
    )

    node = AStarPlanner()

    try:

        rclpy.spin(
            node
        )

    except KeyboardInterrupt:

        pass

    finally:

        node.destroy_node()

        if rclpy.ok():

            rclpy.shutdown()


if __name__ == '__main__':

    main()

File: src/planning/src/a_star_planner_node.cpp
// Fast A* global planner for the DIY Challenge robot.
//
// Nav2 GlobalPlanner plugin implementation. See a_star_planner_node.hpp.

#include "planning/a_star_planner_node.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <queue>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>

namespace diy_astar_planner
{

void AStarPlanner::configure(
  const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
  std::string name,
  std::shared_ptr<tf2_ros::Buffer> tf,
  std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros)
{
  node_ = parent;
  name_ = name;
  tf_ = tf;
  costmap_ros_ = costmap_ros;
  costmap_ = costmap_ros->getCostmap();
  global_frame_ = costmap_ros->getGlobalFrameID();

  auto node = node_.lock();
  logger_ = node->get_logger();
  clock_ = node->get_clock();

  nav2_util::declare_parameter_if_not_declared(
    node, name_ + ".robot_clearance", rclcpp::ParameterValue(robot_clearance_));
  nav2_util::declare_parameter_if_not_declared(
    node, name_ + ".occupied_threshold", rclcpp::ParameterValue(occupied_threshold_));
  nav2_util::declare_parameter_if_not_declared(
    node, name_ + ".unknown_is_occupied", rclcpp::ParameterValue(unknown_is_occupied_));
  nav2_util::declare_parameter_if_not_declared(
    node, name_ + ".visited_publish_interval",
    rclcpp::ParameterValue(visited_publish_interval_));

  node->get_parameter(name_ + ".robot_clearance", robot_clearance_);
  node->get_parameter(name_ + ".occupied_threshold", occupied_threshold_);
  node->get_parameter(name_ + ".unknown_is_occupied", unknown_is_occupied_);
  node->get_parameter(name_ + ".visited_publish_interval", visited_publish_interval_);

  if (visited_publish_interval_ < 1) {
    visited_publish_interval_ = 1;
  }

  visited_map_pub_ = node->create_publisher<nav_msgs::msg::OccupancyGrid>(
    name_ + "/visited_map", rclcpp::QoS(1).transient_local());

  RCLCPP_INFO(logger_, "Configured A* planner \"%s\"", name_.c_str());
  RCLCPP_INFO(logger_, "  robot_clearance = %.2f m", robot_clearance_);
}

void AStarPlanner::cleanup()
{
  RCLCPP_INFO(logger_, "Cleaning up A* planner \"%s\"", name_.c_str());
  visited_map_pub_.reset();
  safe_grid_.clear();
}

void AStarPlanner::activate()
{
  RCLCPP_INFO(logger_, "Activating A* planner \"%s\"", name_.c_str());
  if (visited_map_pub_) {
    visited_map_pub_->on_activate();
  }
}

void AStarPlanner::deactivate()
{
  RCLCPP_INFO(logger_, "Deactivating A* planner \"%s\"", name_.c_str());
  if (visited_map_pub_) {
    visited_map_pub_->on_deactivate();
  }
}

// ================================================================
// Precompute inflated obstacle map from the live costmap
// ================================================================

void AStarPlanner::buildSafeGrid()
{
  const int width = static_cast<int>(costmap_->getSizeInCellsX());
  const int height = static_cast<int>(costmap_->getSizeInCellsY());
  const size_t total_cells = static_cast<size_t>(width) * height;

  clearance_cells_ = static_cast<int>(
    std::ceil(robot_clearance_ / costmap_->getResolution()));

  safe_grid_.assign(total_cells, true);

  std::vector<std::pair<int, int>> blocked_cells;

  for (int y = 0; y < height; ++y) {
    const int row_offset = y * width;
    for (int x = 0; x < width; ++x) {
      const unsigned char value = costmap_->getCost(x, y);

      bool blocked = false;
      if (value == nav2_costmap_2d::NO_INFORMATION) {
        if (unknown_is_occupied_) {
          blocked = true;
        }
      } else if (value >= occupied_threshold_) {
        blocked = true;
      }

      if (blocked) {
        safe_grid_[row_offset + x] = false;
        blocked_cells.emplace_back(x, y);
      }
    }
  }

  // Precompute circular inflation offsets once.
  const int radius = clearance_cells_;
  const int radius_squared = radius * radius;
  std::vector<std::pair<int, int>> inflation_offsets;

  for (int dy = -radius; dy <= radius; ++dy) {
    for (int dx = -radius; dx <= radius; ++dx) {
      if (dx * dx + dy * dy <= radius_squared) {
        inflation_offsets.emplace_back(dx, dy);
      }
    }
  }

  // Inflate every blocked cell.
  for (const auto & obstacle : blocked_cells) {
    for (const auto & offset : inflation_offsets) {
      const int nx = obstacle.first + offset.first;
      const int ny = obstacle.second + offset.second;

      if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
        safe_grid_[ny * width + nx] = false;
      }
    }
  }

  // Treat borders inside clearance radius as unsafe (outside-map space is blocked).
  if (radius > 0) {
    for (int y = 0; y < height; ++y) {
      for (int x = 0; x < width; ++x) {
        if (x < radius || y < radius || x >= width - radius || y >= height - radius) {
          safe_grid_[y * width + x] = false;
        }
      }
    }
  }
}

// ================================================================
// Fast A*
// ================================================================

nav_msgs::msg::Path AStarPlanner::createPlan(
  const geometry_msgs::msg::PoseStamped & start, const geometry_msgs::msg::PoseStamped & goal)
{
  nav_msgs::msg::Path path;
  path.header.frame_id = global_frame_;
  path.header.stamp = clock_->now();

  if (start.header.frame_id != global_frame_ || goal.header.frame_id != global_frame_) {
    RCLCPP_ERROR(
      logger_, "Start/goal frame must be '%s' (got '%s' / '%s')",
      global_frame_.c_str(), start.header.frame_id.c_str(), goal.header.frame_id.c_str());
    return path;
  }

  buildSafeGrid();

  visited_map_.header.frame_id = global_frame_;
  visited_map_.info.width = costmap_->getSizeInCellsX();
  visited_map_.info.height = costmap_->getSizeInCellsY();
  visited_map_.info.resolution = costmap_->getResolution();
  visited_map_.info.origin.position.x = costmap_->getOriginX();
  visited_map_.info.origin.position.y = costmap_->getOriginY();
  visited_map_.data.assign(safe_grid_.size(), 0);

  const GraphNode start_node = worldToGrid(start.pose);
  const GraphNode goal_node = worldToGrid(goal.pose);

  if (!poseOnMap(start_node)) {
    RCLCPP_ERROR(logger_, "Start outside map");
    return path;
  }
  if (!poseOnMap(goal_node)) {
    RCLCPP_ERROR(logger_, "Goal outside map");
    return path;
  }
  if (!isSafe(start_node)) {
    RCLCPP_ERROR(logger_, "Start is inside inflated obstacle region");
    return path;
  }
  if (!isSafe(goal_node)) {
    RCLCPP_ERROR(logger_, "Goal is inside inflated obstacle region");
    return path;
  }

  static const std::vector<std::tuple<int, int, double>> directions = {
    {-1, 0, 1.0}, {1, 0, 1.0}, {0, -1, 1.0}, {0, 1, 1.0},
    {-1, -1, std::sqrt(2.0)}, {-1, 1, std::sqrt(2.0)},
    {1, -1, std::sqrt(2.0)}, {1, 1, std::sqrt(2.0)},
  };

  const int width = static_cast<int>(costmap_->getSizeInCellsX());

  const auto key_of = [width](int x, int y) {
      return static_cast<int64_t>(y) * width + x;
    };

  const int64_t start_key = key_of(start_node.x, start_node.y);
  const int64_t goal_key = key_of(goal_node.x, goal_node.y);

  std::priority_queue<OpenEntry, std::vector<OpenEntry>, OpenEntryGreater> open_heap;

  std::unordered_map<int64_t, double> g_score;
  std::unordered_map<int64_t, int64_t> came_from;
  std::unordered_set<int64_t> closed_set;

  g_score[start_key] = 0.0;

  uint64_t heap_counter = 0;
  const double start_h = octileXY(start_node.x, start_node.y, goal_node.x, goal_node.y);
  open_heap.push({start_h, heap_counter, start_node.x, start_node.y});

  int expanded_count = 0;
  bool goal_reached = false;

  while (!open_heap.empty() && rclcpp::ok()) {
    const OpenEntry current = open_heap.top();
    open_heap.pop();

    const int64_t current_key = key_of(current.x, current.y);

    if (closed_set.count(current_key)) {
      continue;
    }
    closed_set.insert(current_key);
    ++expanded_count;

    visited_map_.data[current.y * width + current.x] = 50;
    if (expanded_count % visited_publish_interval_ == 0) {
      publishVisitedMap();
    }

    if (current_key == goal_key) {
      goal_reached = true;
      break;
    }

    const double current_g = g_score[current_key];

    for (const auto & direction : directions) {
      const int dx = std::get<0>(direction);
      const int dy = std::get<1>(direction);
      const double movement_cost = std::get<2>(direction);

      const GraphNode neighbor{current.x + dx, current.y + dy};

      if (!poseOnMap(neighbor) || !isSafe(neighbor)) {
        continue;
      }

      // Diagonal corner cutting prevention.
      if (dx != 0 && dy != 0) {
        const GraphNode horizontal{current.x + dx, current.y};
        const GraphNode vertical{current.x, current.y + dy};
        if (!isSafe(horizontal) || !isSafe(vertical)) {
          continue;
        }
      }

      const int64_t neighbor_key = key_of(neighbor.x, neighbor.y);
      const double tentative_g = current_g + movement_cost;

      const auto existing_it = g_score.find(neighbor_key);
      const double existing_g =
        existing_it != g_score.end() ? existing_it->second : std::numeric_limits<double>::infinity();

      if (tentative_g >= existing_g) {
        continue;
      }

      came_from[neighbor_key] = current_key;
      g_score[neighbor_key] = tentative_g;

      const double heuristic = octileXY(neighbor.x, neighbor.y, goal_node.x, goal_node.y);
      const double f_score = tentative_g + heuristic;

      ++heap_counter;
      open_heap.push({f_score, heap_counter, neighbor.x, neighbor.y});
    }
  }

  publishVisitedMap();
  RCLCPP_INFO(logger_, "A* expanded %d cells", expanded_count);

  if (!goal_reached) {
    return path;
  }

  // Reconstruct path.
  std::vector<int64_t> cells;
  int64_t current_key = goal_key;
  cells.push_back(current_key);

  while (current_key != start_key) {
    const auto it = came_from.find(current_key);
    if (it == came_from.end()) {
      RCLCPP_ERROR(logger_, "Broken A* parent chain");
      return nav_msgs::msg::Path();
    }
    current_key = it->second;
    cells.push_back(current_key);
  }

  std::reverse(cells.begin(), cells.end());

  for (const int64_t key : cells) {
    const int x = static_cast<int>(key % width);
    const int y = static_cast<int>(key / width);

    geometry_msgs::msg::PoseStamped pose_stamped;
    pose_stamped.header = path.header;
    pose_stamped.pose = gridToWorldXY(x, y);
    path.poses.push_back(pose_stamped);
  }

  return path;
}

// ================================================================
// FAST safety lookup
// ================================================================

bool AStarPlanner::isSafe(const GraphNode & node) const
{
  if (!poseOnMap(node)) {
    return false;
  }
  const size_t index =
    static_cast<size_t>(node.y) * costmap_->getSizeInCellsX() + static_cast<size_t>(node.x);
  return safe_grid_[index];
}

// ================================================================
// Visited map
// ================================================================

void AStarPlanner::publishVisitedMap()
{
  if (!visited_map_pub_ || !visited_map_pub_->is_activated()) {
    return;
  }

  visited_map_.header.stamp = clock_->now();
  visited_map_pub_->publish(visited_map_);
}

// ================================================================
// Heuristic
// ================================================================

double AStarPlanner::octileXY(int x1, int y1, int x2, int y2)
{
  const int dx = std::abs(x1 - x2);
  const int dy = std::abs(y1 - y2);
  const int minimum = std::min(dx, dy);
  const int maximum = std::max(dx, dy);
  return maximum + (std::sqrt(2.0) - 1.0) * minimum;
}

// ================================================================
// Map helpers
// ================================================================

bool AStarPlanner::poseOnMap(const GraphNode & node) const
{
  return node.x >= 0 && node.x < static_cast<int>(costmap_->getSizeInCellsX()) &&
         node.y >= 0 && node.y < static_cast<int>(costmap_->getSizeInCellsY());
}

GraphNode AStarPlanner::worldToGrid(const geometry_msgs::msg::Pose & pose) const
{
  GraphNode node;
  node.x = static_cast<int>(std::floor(
      (pose.position.x - costmap_->getOriginX()) / costmap_->getResolution()));
  node.y = static_cast<int>(std::floor(
      (pose.position.y - costmap_->getOriginY()) / costmap_->getResolution()));
  return node;
}

geometry_msgs::msg::Pose AStarPlanner::gridToWorldXY(int x, int y) const
{
  geometry_msgs::msg::Pose pose;
  pose.position.x = (x + 0.5) * costmap_->getResolution() + costmap_->getOriginX();
  pose.position.y = (y + 0.5) * costmap_->getResolution() + costmap_->getOriginY();
  pose.position.z = 0.0;
  pose.orientation.x = 0.0;
  pose.orientation.y = 0.0;
  pose.orientation.z = 0.0;
  pose.orientation.w = 1.0;
  return pose;
}

}  // namespace diy_astar_planner

#include "pluginlib/class_list_macros.hpp"
PLUGINLIB_EXPORT_CLASS(diy_astar_planner::AStarPlanner, nav2_core::GlobalPlanner)
